### ISSUE TITLE
Add Style/It cop

### ISSUE THREAD

**Issue by bkuhlmann** [rubocop/rubocop#13597]
## Is your feature request related to a problem? Please describe.

This is a new suggestion related to the upcoming Ruby 3.4.0 release where we'll be able to use `it` instead of numbered parameters for blocks. Example:

``` ruby
# No
path.rewrite { |body| new_content + body }
path.rewrite { new_content + _1 }

# Yes
path.rewrite { new_content + it }
```

## Describe the solution you'd like

Would it make sense to have a `Style/It` cop that could detect the above situation and auto-correct accordingly (especially when a block uses only one parameter)? I'm also not sure if `Style/It` is the right name for this but might be nice to have?

## Describe alternatives you've considered

The alternative would be to keep with the status quo.

## Additional context

N/A

**Comment by koic** [comment]
The handling of the `it` parameter's style is not straightforward. For example, options such as the following can be considered:

```ruby
@example EnforcedStyle: allow_named_parameter (default)
  # bad
  block { do_something(_1) }

  # good
  block { |param| do_something(param) }
  block { |it| do_something(it) }

@example EnforcedStyle: always
  # bad
  block { do_something(_1) }
  block { |param| do_something(param) }

  # good
  block { |it| do_something(it) }

@example EnforcedStyle: disallow
  # bad
  block { |it| do_something(it) }

  # good
  block { do_something(_1) }
  block { |param| do_something(param) }
```

It is important that the option which allows named parameters, such as `it`, is the default.  
Additionally, the following points need to be considered:

- Designing support for the `it` parameter syntax as part of the Parser gem interface and adding it to both Prism and RuboCop AST.
- Ensuring it does not conflict with `Style/NumberedParameters`.

The option name `EnforcedStyle: allow_named_parameter` could be further refined. And, the `EnforcedStyle: allow_named_parameter` option effectively conflicts with the role of `Style/NumberedParameters`, which may suggest that it should be disabled by default, requiring further design considerations.

**Comment by bkuhlmann** [comment]
Ah, apologies. I made this confusing by blurring `Style/It` with `Style/NumberedParameters`. To your point, I think (and correct me if I'm wrong), that `Style/NumberedParameters` can remain separate and without further modification if the focus is purely on `Style/It`. Given that assumption, could we do the following by building upon your example:

```
Style/It

@example EnforcedStyle: always
  # bad
  block { |param| do_something param }
  block { |it| do_something it }

  # good
  block { do_something it }

@example EnforcedStyle: disallow
  # bad
  block { do_something it }

  # good
  block { |param| do_something param }
  block { do_something it }
```

I suppose, worst case scenario, is you could see a RuboCop error for both `Style/It` and `Style/NumberedParameters` when both are enabled/enforced and the following code is used:

``` ruby
block { |param| do_something param }
block { |it| do_something it }
```

I would suspect, in practice, people would disable `Style/NumberedParameters` in favor of `Style/It`?

**Comment by Darhazer** [comment]
I'd say we need something for being the only block parameter
e.g. force usage of _1, it or named parameter
force it to be _1 if there are multiple anonymous parameters
or always force named parameters where they are multiples
Perhaps Style/AnonymousParameter
which force (configurable) `it` when it's the only named parameter (not to mix `it` and `_2`) and perhaps which force usage of `it` when it's the only parameter in one line block? We should carefully design the configuration of the cop so it supports wide range of options
E.g. I'd like to
* have `it` if I'm using anonymous parameters and it's the only one
* have named parameters if there is more than 1
* allow a named parameter even if it's the only one
but alternative configuration would be
* have `it` if I'm using anonymous parameters and it's the only one
* have _1 if there is _2, etc.
* force `it` or `_1` if there is one line block
and even more someone might still force `_1` as a style to keep consistency with the second case

**Comment by bkuhlmann** [comment]
Yeah, I like the idea of having a single `Style/AnonymousBlockParameters` for a consolidated configuration of numbered parameters and `it` because they are more intertwined now. I also like the suggestions above. A couple things I'd want to add are:

- Add an option which allows both `it` and named parameters to be disabled entirely for folks that don't want this feature.
- Allow `it` if only one parameter is used and `it` is not disabled (same applies for numbered parameters as you describe).
- Allow consistent style, as you stated, where you use `it` if one, numbered parameters if many, or none/either if all or one or the other is disabled.

There is another situation I ran into recently with the new `it` syntax and it has to do with pipelines. Example:

``` ruby
# Yes
def update_build_options value
  to_h.select { |key, _value| key.start_with? "build_" }
      .transform_values { value }
      .then { |attributes| dup.merge!(**attributes, build_minimum: !value) }
      .freeze
end

# No    
def update_build_options value
  to_h.select { it.start_with? "build_" }
      .transform_values { value }
      .then { dup.merge!(**it, build_minimum: !value) }
      .freeze
end
```

In the *No* situation, I find readability degraded because when you see `it` you have a slight pause of "wait, what is `it` again?" and look up one line versus the *Yes* situation which clearly provides useful naming for improved readability. So what I'm saying is, there might need to be an option to disable use of `it` and numbered parameters in pipelines since I feel like `it` and numbered parameters are most useful for one-liners.

**Comment by dvandersluis** [comment]
> Add an option which allows both it and **named parameters** to be disabled entirely for folks that don't want this feature.

@bkuhlmann do you mean numbered parameters here?

**Comment by dvandersluis** [comment]
I think there is a web of complexity here, because for instance, a consistent style for chained anonymous blocks might be desirable, ie:

```ruby
# inconsistent
foo.bar { _1 + _2 }
   .baz { it }

# consistent
foo.bar { _1 + _2 }
   .baz { _1 }
```

That being said I think it does make sense to manage everything related to anonymous block parameters in a single cop, even if that means retiring `Style/NumberedParameters`.

> @Darhazer: E.g. I'd like to
> *    have it if I'm using anonymous parameters and it's the only one
> *    have named parameters if there is more than 1
> *    allow a named parameter even if it's the only one

This is definitely the style I'd want to use personally as well, I want `it` in place of `_1` for a singular argument, but never `_2`.

**Comment by dvandersluis** [comment]
Definitely not implying we should cover every possibility, just suggesting there are a lot of potential styles that someone could want.
